packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -235,128 +163,168 @@ fn read_file(path: &str) -> Result<String> {
-    let mut try_catch = v8::TryCatch::new(scope);
-    let script = v8::Script::compile(&mut try_catch, js_source_code, None).ok_or_else(|| {
-        let exception = try_catch.exception().unwrap();
-        let exception_string = exception
-            .to_string(&mut try_catch)
-            .unwrap()
-            .to_rust_string_lossy(&mut try_catch);
-
-        Error::Eval(path.into(), format!("Compile error: {exception_string}"))
-    })?;
-
-    script.run(&mut try_catch).ok_or_else(|| {
-        let exception = try_catch.exception().unwrap();
-        let exception_string = exception
-            .to_string(&mut try_catch)
-            .unwrap()
-            .to_rust_string_lossy(&mut try_catch);
-
-        Error::Eval(path.into(), format!("Execution error: {exception_string}"))
+    let mut try_catch_scope = v8::TryCatch::new(scope);
+    let script =
+        v8::Script::compile(&mut try_catch_scope, js_source_code, None).ok_or_else(|| {
+            let exception = read_exception(&mut try_catch_scope).unwrap();
+
+            Error::Eval(path.into(), format!("Compile error: {exception}"))
+        })?;
+
+    script.run(&mut try_catch_scope).ok_or_else(|| {
+        let exception = read_exception(&mut try_catch_scope).unwrap();
+
+        Error::Eval(path.into(), format!("Execution error: {exception}"))
     })
 }
 
-fn import_file<'s>(
-    scope: &mut v8::HandleScope<'s>,
-    context: v8::Local<'s, v8::Context>,
-    path: &str,
-    args: &[Value<'s>],
-) -> Result<Value<'s>> {
-    let v = eval_file(scope, path)?;
-    let f: Function<'_> = v.try_into().map_err(|err| {
-        Error::FileImport(
-            path.into(),
-            format!("Could not convert file into Function: {err}"),
-        )
-    })?;
+/// Cache module to not evaluate them twice
+struct ModuleMap {
+    path_to_module: HashMap<String, v8::Global<v8::Module>>,
+}
+
+impl ModuleMap {
+    fn new() -> Self {
+        Self {
+            path_to_module: HashMap::new(),
+        }
+    }
+
+    fn import_module<'s>(
+        scope: &mut v8::HandleScope<'s>,
+        path: &str,
+    ) -> Result<v8::Local<'s, v8::Module>> {
+        let module_map = scope
+            .get_slot::<Rc<RefCell<ModuleMap>>>()
+            .expect("ModuleMap is not present in isolate slots")
+            .clone();
+
+        if let Some(module) = module_map.borrow().path_to_module.get(path) {
+            return Ok(v8::Local::new(scope, module));
+        }
+
+        let source_code = read_file(path)
+            .map_err(|err| Error::MissingJavascriptImport(path.to_string(), format!("{err}")))?;
+        let js_source_code = new_js_string(scope, &source_code);
+        let js_path = new_js_string(scope, path);
+        let source = v8::script_compiler::Source::new(
+            js_source_code,
+            Some(&v8::ScriptOrigin::new(
+                scope,
+                js_path.into(),
+                0,
+                0,
+                false,
+                0,
```

Comment:
Do you have an idea what the `script_id` is?

Reply:
No, I found [this in deno](https://github.com/denoland/deno/blob/a64e63c3614b98aa2b51fb6b7ef4e30251e03111/core/bindings.rs#L120-L156) which doesn't help that much.

Reply:
A few google results lead me to [chromedevtools.github.io](https://chromedevtools.github.io/devtools-protocol/v8/Runtime/#type-ScriptId), which confirms my thoughts: it's a unique id for a script. We might want to use this ability later, could you add a comment as it's not documented in `rusty_v8`?

Reply:
I didn't put it in the comment since I'm not 100% sure but since we are creating modules and not scripts, this id might not be used in this specific case.
Added comment in https://github.com/hashintel/hash/pull/493/commits/5a7e87bd51e48163e02a324b310a9621d9349677